pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -1392,36 +1430,28 @@ func (p *PriorityQueue) Close() {
 
 // DeleteNominatedPodIfExists deletes <pod> from nominatedPods.
 func (npm *nominator) DeleteNominatedPodIfExists(pod *v1.Pod) {
-	npm.lock.Lock()
-	npm.deleteNominatedPodIfExistsUnlocked(pod)
-	npm.lock.Unlock()
-}
-
-func (npm *nominator) deleteNominatedPodIfExistsUnlocked(pod *v1.Pod) {
+	npm.nominatorLock.Lock()
 	npm.delete(pod)
+	npm.nominatorLock.Unlock()
 }
 
 // AddNominatedPod adds a pod to the nominated pods of the given node.
 // This is called during the preemption process after a node is nominated to run
 // the pod. We update the structure before sending a request to update the pod
 // object to avoid races with the following scheduling cycles.
 func (npm *nominator) AddNominatedPod(logger klog.Logger, pi *framework.PodInfo, nominatingInfo *framework.NominatingInfo) {
-	npm.lock.Lock()
+	npm.nominatorLock.Lock()
 	npm.addNominatedPodUnlocked(logger, pi, nominatingInfo)
-	npm.lock.Unlock()
+	npm.nominatorLock.Unlock()
 }
 
 // NominatedPodsForNode returns a copy of pods that are nominated to run on the given node,
 // but they are waiting for other pods to be removed from the node.
 func (npm *nominator) NominatedPodsForNode(nodeName string) []*framework.PodInfo {
-	npm.lock.RLock()
-	defer npm.lock.RUnlock()
-	// Make a copy of the nominated Pods so the caller can mutate safely.
-	pods := make([]*framework.PodInfo, len(npm.nominatedPods[nodeName]))
-	for i := 0; i < len(pods); i++ {
-		pods[i] = npm.nominatedPods[nodeName][i].DeepCopy()
-	}
-	return pods
+	npm.nominatorLock.RLock()
+	nominatedPods := append(make([]NominatedPod, 0, len(npm.nominatedPods[nodeName])), npm.nominatedPods[nodeName]...)
+	npm.nominatorLock.RUnlock()
+	return npm.nominatedPodsToInfo(nominatedPods)
```

Comment:
(I know I suggest adding the same comment everywhere, but it's worthy, it's critically important.)

```suggestion
	// Note that nominatedPodsToInfo takes SchedulingQueue's qlock inside.
	return npm.nominatedPodsToInfo(nominatedPods)
```